nix_update/version/github.py

```diff
@@ -34,3 +36,100 @@ def fetch_github_versions(url: ParseResult) -> List[Version]:
     tree = ET.fromstring(resp.read())
     releases = tree.findall(".//{http://www.w3.org/2005/Atom}entry")
     return [version_from_entry(x) for x in releases]
+
+
+def init_git(git_dir: str, url: str) -> None:
+    run(["git", "init", "-q", git_dir], stdout=None)
+    run(
+        [
+            "git",
+            "-C",
+            git_dir,
+            "remote",
+            "add",
+            "origin",
+            url,
+        ],
+        stdout=None,
+    )
+
+
+def get_latest_snapshot_from_git(
+    git_dir: str, branch: str, before: str, increment: int = 20, initial: int = 20
+) -> Version:
+    run(
+        [
+            "git",
+            "-C",
+            git_dir,
+            "fetch",
+            f"--depth={initial + increment}",
+            "origin",
+            branch,
+        ],
+        stdout=None,
+    )
+    while True:
+        log = run(
+            [
+                "git",
+                "-C",
+                git_dir,
+                "log",
+                "--format=%cd %H",
+                "--date=format-local:%F",
+                f"--before={before}T00:00:00Z",
+                "--max-count=1",
+                f"origin/{branch}",
+            ],
+            extra_env={"TZ": "UTC"},
+        ).stdout.strip()
+        if log:
+            date, commit = log.split(maxsplit=1)
+            return Version(f"unstable-{date}", rev=commit)
+        run(
+            [
+                "git",
+                "-C",
+                git_dir,
+                "fetch",
+                f"--deepen={increment}",
+                "origin",
+                branch,
+            ],
+            stdout=None,
+        )
+
+
+def fetch_github_snapshots(url: ParseResult, branch: str) -> List[Version]:
+    if url.netloc != "github.com":
+        return []
+    parts = url.path.split("/")
+    owner, repo = parts[1], parts[2]
+    repo = re.sub(r"\.git$", "", repo)
+    feed_url = f"https://github.com/{owner}/{repo}/commits/{branch}.atom"
+    info(f"fetch {feed_url}")
+    resp = urllib.request.urlopen(feed_url)
+    tree = ET.fromstring(resp.read())
+    commits = tree.findall(".//{http://www.w3.org/2005/Atom}entry")
+
+    before = datetime.utcnow().strftime("%Y-%m-%d")
+    snapshots = []
+    for entry in commits:
+        link = entry.find("{http://www.w3.org/2005/Atom}link")
+        updated = entry.find("{http://www.w3.org/2005/Atom}updated")
+        if link is None or updated is None or updated.text is None:
+            continue
+        url = urlparse(link.attrib["href"])
+        commit = url.path.rsplit("/", maxsplit=1)[-1]
+        date = updated.text.split("T", maxsplit=1)[0]
+        if date < before:
+            snapshots.append(Version(f"unstable-{date}", rev=commit))
+            before = date
```

Comment:
The API only gives us the commit the branch points to, whereas the feed gives us a couple more commits and therefore a better chance to find a day's last commit.

Reply:
Why do we need a days last commit? I thought we always want to update to the latest commit available.

Reply:
I would do it this way so every version is unique. Otherwise multiple commits of a day would have exactly the same version and actual updates wouldn't change the version.

Reply:
I'm thinking about adding a new sub-version like `unstable-YYYY-MM-DD.n` where `n` is n-th commit of the day or using the total number of commits like <https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=invidious-git#n24>.

Reply:
> I would do it this way so every version is unique. Otherwise multiple commits of a day would have exactly the same version and actual updates wouldn't change the version.

That's how nixpkgs currently works.

> I'm thinking about adding a new sub-version like `unstable-YYYY-MM-DD.n` where `n` is n-th commit of the day or using the total number of commits like [aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=invidious-git#n24](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=invidious-git#n24).

The version format is getting a revamp in the future. Please don't introduce anything new and this shouldn't be accepted in nixpkgs anyway.

Reply:
I added a new version `--before[=YYYY-MM-DD]` that tells nix-update to only pick commits the specified date. if `YYYY-MM-DD` is omitted only commits before today are used.

Reply:
I think using the atom feed is better because you cannot run into API limits this way. I prefer the current approach actually.